src/hsmm_mvpy/models.py

```diff
@@ -1952,75 +2211,195 @@ def fit(self, step=1, verbose=True, end=None, trace=False, fix_iter=False, max_i
+                mags[:n_events] = solutions.magnitudes.values
+                pars[:n_events+1] = solutions.parameters.values
+                #locations[:n_events+1] = solutions.locations.values
+                
+                #store solution
+                estimates.append(solutions)
+
+                #search for an additional event, starting again at sample 1 from prev event,
+                #or next sample if by_sample
                 n_events += 1
-                j = 1
-                if verbose:
-                    print(f'Transition event {n_events} found around sample {int(np.round(self.scale_to_mean(np.sum(pars[:n_events,1]), self.shape)))}')
-            j += 1
-            i += 1
-            #New parameter proposition
-            pars_prop = pars[:n_events+2].copy()
-            pars_prop[n_events,1] = self.mean_to_scale(step*j, self.shape)
-            time = int(np.round(self.scale_to_mean(np.sum(pars[:n_events,1]), self.shape)))
+                if by_sample:
+                    j += 1
+                    time = j
+                else:
+                    j = 1
+                    time = sol_sample_new_event + j
+
+
+                #Diagnostic plot
+                if diagnostic:
+                    color = next(cycol)
+                    plt.plot(solutions.traces.T, c=color, label=f'n-events {n_events-1}')
+                if verbose: 
+                    print(f'Transition event {n_events-1} found around sample {sol_sample_new_event}')
+
+
+            else: #reject solution, search on
+                prev_sample = int(np.round(self.scale_to_mean(np.sum(pars[:n_events-1,1]), self.shape)))
+
+                if not by_sample: #find furthest explored param. Note: this also work by_sample, just a tiny bit faster this way
+                    max_scale = np.max([np.sum(x[:n_events,1]) for x in solutions.param_dev.values])
+                    max_sample = int(np.round(self.scale_to_mean(max_scale, self.shape)))
+                    j = np.max([max_sample - prev_sample + 1, j + 1]) #either ffwd to furthest explored sample or add 1 to j
+                    time = prev_sample + j
+                else:
+                    j += 1
+                    time = j
+
             pbar.update(int(np.rint(time-prev_time)))
-        pbar.update(int(np.round(np.rint(end))-np.rint(time)))
+
+        #done estimating
+
+        n_events = n_events-1
+        if verbose:
+            print()
+            print('All events found, refitting final combination.')
         if diagnostic:
             plt.ylabel('Log-likelihood')
             plt.xlabel('EM iteration')
             plt.legend()
             plt.show()
         mags = mags[:n_events, :]
         pars = pars[:n_events+1, :]
-        if n_events > 0: 
+        locs = locations[:n_events+1]
+        if n_events > 0:
+            #fit = self.fit_single(n_events, parameters=pars, magnitudes=mags, locations=locs, verbose=verbose)
```

Comment:
How come that you don't use locations here?

Reply:
sometimes it has build up high, unnecessary locations during estimation (unnecessary because the events have moved far apart). Here I give it one last chance to discover smaller locations, which can make more flexible if the solutions are used in additional analyses. I've gone back and forth on this, but the is seemed to work best in the end.